Add Classic Text control, including list controls

kienstra opened the pull request:
### Update: this has changed, please see this [comment below](https://github.com/getblocklab/block-lab/pull/432#issuecomment-538701506).
* This begins to add list buttons to the Rich Text controls
* Still, it might not be possible to get this to work as we'd expect. For example, it looks like it's not possible to have the list buttons next to the bold and italic buttons:

<img width="679" alt="rich-text-control" src="https://user-images.githubusercontent.com/4063887/65123699-6cdb4f80-d9b9-11e9-8ea7-8174f6f34b19.png">

Also, it might be really hard to make only part of the text a list.

For example, in the Core List block, it simply makes its entire `<RichText>` component a `<ul>` or `<ol>`. I'm not sure if the `<RichText>` has a good way to make only part of the text into a `<ul>` or `<ol>`.

Closes #405 (if this works)

kienstra committed:
Begin adding list controls to the Rich Text control

Still, it might not be possible to get
this to work,
as the buttons appear in the popover.

kienstra committed:
Remove a needless </RichText>

It's possible to simply close this
with />.

kienstra committed:
Remove duplicate wp.blockEditor destructuring

There was already a const declaration withwp.blockEditor
so simply use that .

kienstra committed:
Explore using the Core Classic Editor's Edit component

This uses TinyMCE, but this looks much simpler
than using the RichText component.

kienstra commented:
**Use TinyMCE?**

Hi @lukecarbis,
What do you think about https://github.com/getblocklab/block-lab/pull/432/commits/0f48d3d0e411fbd5456e2471f2452af5785ebac6, which uses TinyMCE instead of the `<RichText>` component? 

The list buttons work well so far, and it even has alignment controls. Though this PR could use some more polish.

So far, it looks to be backwards-compatible with the earlier Rich Text control. It's mainly forked from the [Core Classic block](https://github.com/WordPress/gutenberg/blob/416c9bc9eaef6e6bceea923b6f36642746c01aba/packages/block-library/src/classic/edit.js).

Also, it's much easier to use than `<RichText>` so far. 

<img width="1065" alt="rich-text-tiny" src="https://user-images.githubusercontent.com/4063887/65177628-ae511680-da1c-11e9-8312-131e77e9aa26.png">

kienstra commented:
**Lists**

Lists are working, but they could use better styling:

<img width="772" alt="lists-work-here" src="https://user-images.githubusercontent.com/4063887/65178096-a8a80080-da1d-11e9-998e-f774d0159f2c.png">

kienstra commented:
Another bonus is that we could later add controls that the Classic block has, like text color, strikethrough, or horizontal line.

kienstra commented:
If you have time, feel free to commit styling changes on this PR.

kienstra committed:
Add JS DocBlocks to methods, prepare for code review

Also, address an issue
where the content wasn't saving on clicking
somewhere else in the editor.

kienstra committed:
Remove the check for the TinyMCE being empty

This was from entering delete for the
Classic block.
But it doesn't apply here,
as we don't want someone to delete this entirely

kienstra committed:
Fix an issue where typing enter moved the cursor to the beginning

This used to move it to the beginning.
Also, fix an issue where when the image is floated,
the container is too small.

kienstra committed:
Add a strikethrough button for backwards compatibility

There was one in the previous RichText,
though it looks like there isn't simply
a button for code, like there was in the
previous <RichText>.

kienstra committed:
Make the editor look more like the previous implementation

Use the white background color
instead of the darker gray.
Also, use the focus color of blue,
like at least one other field.

kienstra committed:
In the ServerSideRender component, improve <ul> and <ol> display

Before, these overflowed their container (to the left).
So apply 1rem of margin.

kienstra commented:
**Request For Review**

Hi @lukecarbis and @RobStino,
Could you please review this? It adds list and alignment controls to the Rich Text control.

# Features Of New Rich Text

<img width="822" alt="new-showing" src="https://user-images.githubusercontent.com/4063887/65218066-441b8e80-da7b-11e9-9f9e-7fde4aaa3054.png">

As far as I've tested this, it's backwards-compatible with the previous control. But the 'Code' button doesn't look to be available in the new TinyMCE implementation:

<img width="547" alt="code-button" src="https://user-images.githubusercontent.com/4063887/65217798-ade76880-da7a-11e9-9c96-0ab59f0bc092.png">

Still, text that was code-formatted with the previous control still looks the same on loading this new implementation.

# Before

<img width="881" alt="before-rt" src="https://user-images.githubusercontent.com/4063887/65217688-67920980-da7a-11e9-820b-b6969684f6cc.png">

# After 

<img width="858" alt="after-rt" src="https://user-images.githubusercontent.com/4063887/65217696-6d87ea80-da7a-11e9-982a-edae457cc652.png">

This also surfaced a bug with the previous Rich Text implementation, which I think this new version fixes.

Before, if there was an empty line:

<img width="893" alt="previous-rich-text" src="https://user-images.githubusercontent.com/4063887/65217894-e7b86f00-da7a-11e9-8403-73c49926f0e6.png">

...it didn't display in the `<ServerSideRender>` or the front-end:

<img width="879" alt="server-side-render-bad-display" src="https://user-images.githubusercontent.com/4063887/65217917-f868e500-da7a-11e9-97c5-13fedc739cdc.png">

This PR fixes that. That might be considered a backwards-compatibility break, but I think it's more of a bug fix.

lukecarbis reviewed (changes requested):
Looking great Ryan! I can't believe I'm saying this, but I think switching to TinyMCE might actually be a good idea.

Couple of things I noticed:

1. The toolbar isn't displayed until the textarea is focused. After that it remains, even on blur. My suggestion is to make it always visible. Alternatively make it only show when the textarea is focused, and the mouse is moving (hide again when you start typing), similar to the previous implementation of rich text.

![2019-09-23 08-32-10 2019-09-23 08_32_35](https://user-images.githubusercontent.com/1097667/65395352-ed94b400-dddc-11e9-95b6-8581f41afe5e.gif)

2. A few styling suggestions for the toolbar:
- Remove the box-shadow from `.mce-tinymce`
- Center align
- White background
- 2px bottom margin
- 4px border radius on `.mce-toolbar .mce-btn-group .mce-btn`
- Make the buttons match the current rich text style (different hover / active effect)

![Screen Shot 2019-09-23 at 8 47 38 am](https://user-images.githubusercontent.com/1097667/65395496-d2c33f00-ddde-11e9-85b9-ee9a746a1c2b.png)

3. Placeholder text doesn't seem to work.

———

More on the use of TinyMCE. What do you think about this idea:

- We use the `<TinyMCE>` component in our  "Rich Text" control, as per this implementation
- Rather than removing the `<RichText>` component entirely, we switch our current Textarea control to use the `<RichText>` component
- Possibly, we simplify the `<RichText>` component in Textarea to remove the Inline Image

kienstra committed:
Apply suggested styling fixes

This applies many of Luke's styling improvements.

Still, there is work needed on having the toolbar
display on the block first displaying.

kienstra commented:
**Thanks for reviewing this!**

Hi @lukecarbis,
Good point here!

1. Ah, good catch. I'll work on that tomorrow if that's alright. Though it'd be really good to have this PR in a release this week. There's a user that needs this for a project.

2. Thanks, those suggestions are applied in https://github.com/getblocklab/block-lab/pull/432/commits/9714ee781eeb7f1ae3db65730d2190d7cc0f763e:

<img width="778" alt="toolbar-styling" src="https://user-images.githubusercontent.com/4063887/65404782-902e4200-dd9f-11e9-9013-d0a962022177.png">

3. Yeah, good catch. What do you think of removing the placeholder? 

I didn't see a good way to add a placeholder for TinyMCE. There are [workarounds](https://stackoverflow.com/questions/27237876/how-do-i-add-placeholder-text-to-tinymce), like adding a TinyMCE plugin. 

UX is definitely not my speciality, but I'm not sure if a placeholder applies to a full rich text editor.

kienstra commented:
>Rather than removing the <RichText> component entirely, we switch our current Textarea control to use the <RichText> component

Interesting 😄 It's probably possible, we could probably `wpautop()` the previous Textarea value, to make it compatible with a `<RichText>` component.

UX isn't my specialty 😄 But I'm not sure if people would expect bold, italic, and link controls in a Textarea field.

kienstra commented:
Hi @lukecarbis,
You've been really productive today already 😄 

But if you happen to have time and want to dig into it, feel free to work on [point 1 above](https://github.com/getblocklab/block-lab/pull/432#pullrequestreview-291531559) and commit to this branch.

I'll have a pretty busy day tomorrow, and might not be able to completely fix it.

kienstra committed:
Make the toolbar always visible

Remove the conditiona that wrapped this.editor.focus();
Still, I'm not positive why that condition was there.

lukecarbis commented:
@kienstra Thanks for looking at all that. Larger reply still incoming, but just wanted to mention that I noticed that although https://github.com/getblocklab/block-lab/pull/432/commits/effe21d425e1a6b6f2679d57a8bfd4c60ca7ae1c helps keep the toolbar visible after focusing, it's not visible when first adding the block to a post.

lukecarbis commented:
Also, the "selected" state doesn't match the regular <RichText> component. See the "Bold" below.

![Screen Shot 2019-09-24 at 8 12 08 am](https://user-images.githubusercontent.com/1097667/65466761-1e3f2100-dea3-11e9-9231-0e671c8b29bb.png)

kienstra committed:
Use :hover and .is-active styling from RichText component

As Luke mentioned,
this mainly copies the styling from the RichText component.

kienstra commented:
Here's the current display of buttons with `.is-active` and `:hover`:

![active-hov](https://user-images.githubusercontent.com/4063887/65466980-e219ab00-de25-11e9-937b-943d047c01a7.gif)

kienstra committed:
Revert "Make the toolbar always visible"

This reverts commit effe21d425e1a6b6f2679d57a8bfd4c60ca7ae1c.

kienstra commented:
Sorry, I probably won't have more time today to work on this.

kienstra commented:
Hm, I might have made some progress on this.

kienstra commented:
This looks to fix the issue of the toolbar not displaying, but it has its own issues:

```diff
diff --git a/js/blocks/components/tiny-mce.js b/js/blocks/components/tiny-mce.js
index 9923331..38ddc76 100644
--- a/js/blocks/components/tiny-mce.js
+++ b/js/blocks/components/tiny-mce.js
@@ -71,10 +71,12 @@ class TinyMCE extends Component {
                wp.oldEditor.initialize( `editor-${ editorId }`, {
                        tinymce: {
                                ...settings,
-                               inline: true,
                                content_css: false,
                                fixed_toolbar_container: `#toolbar-${ editorId }`,
+                               inline: false,
+                               resize: false,
                                setup: this.onSetup,
+                               statusbar: false,
                                toolbar1: 'bold,italic,bullist,numlist,outdent,indent,alignleft,aligncenter,alignright,link,unlink,wp_add_media,strikethrough',
                                toolbar2: '',
                        },
```

kienstra commented:
Sorry, that's probably all I can work on this today.

lukecarbis commented:
My suggestion at this point is to create a new control to house this `<TinyMCE>` component (maybe called "Classic Text"), and keep the "Rich Text" control as it currently is.

RobStino commented:
Hey all.
Tested this fairly thoroughly and overall I really like this!
I liked how it didn't wrap the output on the frontend with any divs or anything.

Here's the field in the editor:
![image](https://user-images.githubusercontent.com/1578148/65659044-6a3dc180-e06d-11e9-92ef-489f7451ab10.png)
Looked good in the Editor when deselected:
![image](https://user-images.githubusercontent.com/1578148/65659062-79bd0a80-e06d-11e9-9b5a-283432adea03.png)

Looked and behaved as expected on the frontend as well.

My suggestions for improvements (if possible/practical) include:
- Set the default height to 5 rows. Currently, it's only 1 row.
- Include Heading control
- Include code font style control

Possible other controls, although maybe not...
- Font color
- Font size

As far as moving forward I can see how this is experience is what we want, but can't yet deliver in the Rich Text. I think I'm in favour of having both, and then potentially merging/migrating at some point in the future when we can achieve feature parity.

My only concern is that it kind of goes against our "it just works" product philosophy. Having two options that do kinda the same thing doesn't feel great. But, it is what it is. Happy to defer on this one.

RobStino commented:
I also observed the bug where the controls bar doesn't display on initial load.

kienstra committed:
Make toolbar visible on first adding the block, and before focusing

Now, the toolbar should be visible any time the
'Classic Text' block is visible.
Thoguh this could still use more testing.

kienstra committed:
Make the minimum height of the Classic Text block the same as 3 linesS

Now, even if there's no text in it,
it has the height of 3 lines.

kienstra commented:
With https://github.com/getblocklab/block-lab/pull/432/commits/28b88d52e86c725d60381233b8e24e738aebae7a, the `min-height` of this Control is the equivalent of 3 lines, even if there's no text:

<img width="905" alt="min-height-rich-text" src="https://user-images.githubusercontent.com/4063887/66262138-8691da80-e7a0-11e9-8ed1-fc3fe792a0d2.png">

Feel free to change the `min-height`.

Also, it looks like the toolbar now displays on first creating a block, and on focusing the block. It doesn't require clicking in the edit area anymore.

kienstra commented:
Once #444 is merged into `develop`, merging `develop` into this branch should make the failing unit tests above pass.

Update: I applied the same fix in this PR, so it should be good.

kienstra committed:
Make the Classic Text control separate from Rich Text

Before, this had changed the Rich Text
control to use TinyMCE.
But this adds a separate control for the TinyMCE:
Classic Text.
There should be minimal changes to the Rich Text control.

kienstra commented:
With https://github.com/getblocklab/block-lab/pull/432/commits/3cd1f5285ab0e1aa6e11fc8b8c4e7d64e8518aca, there is now a separate control, Classic Text. 

This PR's changes for the Rich Text control are moved into that, and the Rich Text control is mainly unchanged in this PR.

<img width="828" alt="new-ct" src="https://user-images.githubusercontent.com/4063887/66262599-69153e80-e7a9-11e9-88c4-5bd6ea3a5471.png">

This doesn't have a placeholder, though it has help text and a default value.

kienstra committed:
Don't run wpautop() for 'Classic Text' control

It looks like this already adds <p></p> where appropriate,
so this probably isn't needed.

kienstra committed:
Add 'formatselect' to the toolbar

This enables choosing <h1>, <h2>, etc..

and 'preformatted'.

kienstra commented:
There is now a 'formatselect' in the toolbar:

<img width="924" alt="here-is-a-heading" src="https://user-images.githubusercontent.com/4063887/66273546-79292e80-e83a-11e9-92a7-75d1602f0bec.png">

kienstra committed:
Add the Classic Text block to the integration tests

This is pretty simple,
as it should output the same value that's saved.

kienstra committed:
Remove references to the Rich Text component in Classic Text

These are from when they were the same component.
So change these to 'Rich Text'.

kienstra commented on tests/php/unit/helpers/trait-testing-helper.php:
```diff
@@ -42,7 +42,7 @@ public function set_license_validity( $is_valid ) {
 			$transient_value = array(
 				'license' => 'valid',
-				'expires' => date( '+1 month' ),
+				'expires' => date( 'D, d M Y H:i:s', time() + 1000 ),
```
This fixes the failed unit tests. I'm not sure how this worked before 😄

kienstra committed:
Add the 'block-lab' textdomain to __()

This was forked from Gutenberg,
so it'd be best to add out textdomain to this,
in case the source file changes.

kienstra committed:
Correct DocBlocks for 2 classes

These were probably the result of pasting
from the Textarea class.

kienstra commented:
**Request For Review**

Hi @lukecarbis,
Hope you had a great weekend.

Could you please review this Classic Text control?

Like we talked about, this creates a new control, and the Rich Text control is hardly changed in this PR.

<img width="768" alt="classic-text-control" src="https://user-images.githubusercontent.com/4063887/66274187-8d245e80-e841-11e9-8c14-bc6db4593add.png">

Thanks, Luke!

kienstra commented:
**Thanks For Reviewing**

Hi @RobStino,
Sorry for the delay in replying to your [review](https://github.com/getblocklab/block-lab/pull/432#issuecomment-535337069).

>My suggestions for improvements (if possible/practical) include:

> Set the default height to 5 rows. Currently, it's only 1 row.

Good idea. Sorry, I missed your comment, and only made it 3 rows. How is that? It's easy to change here if you'd prefer:

<img width="905" alt="min-height-rich-text" src="https://user-images.githubusercontent.com/4063887/66262138-8691da80-e7a0-11e9-8ed1-fc3fe792a0d2.png">


> Include Heading control

Good suggestion, that's applied [here](https://github.com/getblocklab/block-lab/pull/432#issuecomment-538772431).

> Include code font style control

Sure, that's now included in the [formatselect](https://github.com/getblocklab/block-lab/pull/432#issuecomment-538772431), as 'Preformatted'.

> Possible other controls, although maybe not...
> Font color
> Font size

Sure, any controls in the Core Classic block are available here. Though adding more might require another toolbar row, like the Core Classic block has:

<img width="768" alt="classic-text-control" src="https://user-images.githubusercontent.com/4063887/66274271-50a53280-e842-11e9-815a-60b470d6d660.png">

lukecarbis commented on js/blocks/components/tiny-mce.js:
```diff
@@ -0,0 +1,201 @@
+const { F10 } = wp.keycodes;
+
+/**
+ * Forked from the Core Classic Editor's Edit component.
```
What do we get by forking instead of just including?

kienstra replied:
Yeah, it'd be good to include it somehow.

But I don't think there's a good way to do that.

For example, the Classic block's `index.js` file [imports it](https://github.com/WordPress/gutenberg/blob/416c9bc9eaef6e6bceea923b6f36642746c01aba/packages/block-library/src/classic/index.js#L9):


```js
import edit from './edit';
```

lukecarbis commented on php/post-types/class-block-post.php:
```diff
@@ -41,6 +41,7 @@ class Block_Post extends Component_Abstract {
 	public $pro_controls = array(
 		'repeater',
 		'post',
+		'classic_text',
```
Can we please switch these two lines so that Rich Text appears before Classic?

kienstra replied:
Sure

kienstra replied:
It's moved in https://github.com/getblocklab/block-lab/pull/432/commits/1e6aac76933c3a661c69c73d3bb9c9664b81c269

lukecarbis commented:
@kienstra This is looking great! I left a couple of questions, but at this point I think it's just about getting the styles looking right.

I know we've gone to some effort to get the controls to match the Rich Text, but I think we should actually change that now! :/ Sorry.

Partly, this is due to the width of the toolbar now much wider due to the formatselect.

Also, we were trying to replicate the RichText style earlier because we intended to _replace_ it. Now that we're embracing the "Classic"ness, maybe we should style the toolbar more like this:

<img width="719" alt="Screen Shot 2019-10-07 at 9 31 56 am" src="https://user-images.githubusercontent.com/1097667/66277707-a2eb5f80-e8e5-11e9-99e6-e77407d92d5d.png">

(Not saying we need all these controls, just that our toolbar could look / be styled similarly).

@RobStino , do you have any opinion on this?

kienstra committed:
Move the Classic Text block to after the Rich Text block

In the $pro_controls array,
move the Classic Text block later.

kienstra commented:
Yeah, it'll probably be good to remove the styling that made this look like the `<RichText>` component. It should also make this more maintainable, with less styling.

RobStino commented:
Hey fellas. I agree with Luke. Let's maintain the styling of the Classic block, so to speak. Maintainability it makes sense, and it also differentiates it from the RichText further.

kienstra commented:
Thanks, @RobStino! That sounds good. I'll apply that to this PR in the next few days.

kienstra committed:
Remove all custom toolbar styling except to make it centered and white

In the toolbar,
there was custom styling to match the <RichText> component.
Now that this is a different block,
this probably doesn't need that styling.

kienstra commented:
**Question About New Styling**

Hi @lukecarbis and @RobStino,
What do you think about this new styling for the toolbar? https://github.com/getblocklab/block-lab/pull/432/commits/44e95d20d2c3dd587cade0017ab4a3298f96bdc7 removes all of the custom styling, except the rules that centered it and made it white:

<img width="898" alt="buttons-s" src="https://user-images.githubusercontent.com/4063887/66599936-fe2d8400-eb69-11e9-9f5a-2ec31d987cbc.png">

We could also remove the styling that centers the toolbar and makes it white. I'm not sure if that looks good, but that's definitely not my specialty 😄 

<img width="853" alt="classic-text-distinct-buttons" src="https://user-images.githubusercontent.com/4063887/66599860-d3dbc680-eb69-11e9-9795-fd0dbb20ebb6.png">

lukecarbis committed:
Update styles for the classic text block

lukecarbis commented:
@kienstra – Hope you don't mind, I pushed a commit (https://github.com/getblocklab/block-lab/pull/432/commits/674f66207c5537c0fe7bece9915b18625b3df9c7) which has some style updates. 

Here's what it looks like now:
![Screen Shot 2019-10-13 at 9 00 51 am](https://user-images.githubusercontent.com/1097667/66708696-1c78c700-ed98-11e9-9a09-c9a82d83b1b5.png)

And focused:
![Screen Shot 2019-10-13 at 9 00 37 am](https://user-images.githubusercontent.com/1097667/66708699-213d7b00-ed98-11e9-9be2-792a0033db0f.png)

@RobStino Can I get your sign-off on these designs?

lukecarbis commented:
Found one last bug! Related to the Classic Text stealing focus.

1. Create a new block, which includes a Classic Text field, as well as some other text controls.
2. Make sure the Classic Text field is not the _first_ one.
3. Add the Block to a post
4. Notice that the Classic Text field is focused first when the Block is selected

![2019-10-13 09-05-09 2019-10-13 09_07_30](https://user-images.githubusercontent.com/1097667/66708752-f869b580-ed98-11e9-94e3-0baa987dc3d8.gif)

kienstra commented:
Ah, good catch! I'll probably get to this within a few days.

kienstra commented:
It looks like there might not be an easy fix for the issue with the [Classic editor's toolbar focusing](https://github.com/getblocklab/block-lab/pull/432#issuecomment-541368486):

https://github.com/WordPress/gutenberg/pull/4948

kienstra committed:
Attempt a fix for the Classic Text focus issue

It looks like this fixes the issue of the Classic Text
control getting focus on the block loading,
even if there are other controls above it.
Still, this could use more testing,
and it could have side-effects.

kienstra committed:
Revert "Attempt a fix for the Classic Text focus issue"

This reverts commit 5861a3546be9a41feb7a6d2933dd01c0d6760717.

kienstra commented:
I'll look at this a little more, but there [might not be a fix](https://github.com/WordPress/gutenberg/pull/4948#issuecomment-365242220) for the [Classic editor's toolbar focusing](https://github.com/getblocklab/block-lab/pull/432#issuecomment-541368486):

>I don't really see a way of fixing the focus problem without having to copy over a significant amount of code to the classic block to overwrite the theme. It would be nicer if the inline theme just had a setting to show toolbars all the time instead of just when the editor is focussed.

kienstra commented:
One option would be to only display the toolbar on focusing the TinyMCE. This would at least prevent the [issue with focusing](https://github.com/getblocklab/block-lab/pull/432#issuecomment-541368486).

kwfricke commented:
This is awesome. I CAN'T WAIT for this to be released :D :D

kienstra commented:
Thanks, @kwfricke! Looking forward to getting it to you 😄

kienstra commented:
**Question About Toolbar**

Hi @lukecarbis,
Hope you're doing great. Could you please help with a question?

I couldn't find a fix for the [issue with the Classic Text control's toolbar focusing](https://github.com/getblocklab/block-lab/pull/432#issuecomment-541368486).

It looks like the TinyMCE that this uses requires a focus event for the toolbar to display.

The [workaround in Gutenberg](https://github.com/WordPress/gutenberg/blob/853b3cb100469f9d71bad93bb38ece5210dae0b2/packages/block-library/src/classic/edit.js#L157) seems to work because when you add a Classic block, the TinyMCE is focused. That's pretty much all the block has 😄 

But when you add a Block Lab block that has several fields, including a TinyMCE (Classic Text), it's more complex.

Do any of these options sound acceptable?

1. Only display the toolbar on clicking in (focusing) the Classic Text field (like [how it was in this PR before](https://github.com/getblocklab/block-lab/pull/432#pullrequestreview-291531559))
2. Keep this PR as is, with the known [issue with focusing](https://github.com/getblocklab/block-lab/pull/432#issuecomment-541368486)

Thanks, Luke!

lukecarbis commented:
@kienstra Let's go with option 1: Only display the toolbar on clicking in (focusing) the Classic Text field (like how it was in this PR before).

@kwfricke Sorry this is taking so long! We want to make sure it's right! :)

kienstra commented:
Thanks, Luke! That sounds good.

kienstra committed:
Revert the focus behavior to what it was before

Unfortunately, it doesn't look possible to
have the toolbar display by default.
It will now only display on focusing the editing area.
Also, remove some style rules that looked a little strange
with this new behavior.
Still, more styling work will be probably needed.

kienstra commented:
**Request For Some Styling Work And Final Review**

Hi @lukecarbis,
Thanks, https://github.com/getblocklab/block-lab/pull/432/commits/fd8e87f9166bbe55b66b24aebde75de2c127bc67 makes the toolbar only display on clicking in (focusing) the Classic Text field (like [how it was in this PR before](https://github.com/getblocklab/block-lab/pull/432#pullrequestreview-291531559)).

Your [styling commit](https://github.com/getblocklab/block-lab/pull/432/commits/674f66207c5537c0fe7bece9915b18625b3df9c7) looked great, and unfortunately, that commit I just made to the JavaScript messed it up 😄 

So I made a few edits to `editor.scss`, but could you please work on the styling more? 

<img width="948" alt="class-text-control" src="https://user-images.githubusercontent.com/4063887/67149268-ebeccd80-f26e-11e9-8d13-ff1867e7b128.png">

Thanks, Luke!

kienstra committed:
Merge branch 'develop' into add/rich-text-lists

lukecarbis committed:
Style updates to the classic text toolbar

lukecarbis commented:
@kienstra Styles are in. Is it possible to hide the toolbar when the classic text area loses focus? Or was that one of the limitations of this approach?

lukecarbis commented:
FWIW, the Rich Text toolbar doesn't disappear when it loses focus either. I would normally create a separate issue for this – but if there's a simple win here, may as well include it in this PR along with the Classic Text toolbar.

kienstra commented:
>@kienstra Styles are in. Is it possible to hide the toolbar when the classic text area loses focus? Or was that one of the limitations of this approach?

Hi @lukecarbis, 
It might be possible, but if it's alright, the weekend might be the earliest I could work on this. Would it be acceptable to ship this as is?

kienstra commented:
**Request For Final Review**

Hi @RobStino,
When you have a chance, could you please review this? 

There's a [known issue](https://github.com/getblocklab/block-lab/pull/432#issuecomment-544179824) where the toolbar only displays on clicking the edit area. There doesn't look to be a workaround.

Thanks, Rob!

kienstra commented:
>@kienstra Styles are in. Is it possible to hide the toolbar when the classic text area loses focus? Or was that one of the limitations of this approach?

I'll look at this more this weekend, but I don't think it'll be easy to hide the toolbar. There doesn't look to be an API for the toolbar, it's only shown on focus with a workaround (maybe hack).

lukecarbis commented:
@kienstra As discussed in our last conversation, I'm happy with where is at, and we're all happy to move forward with this PR understanding that there's a limitation with the toolbar not hiding.

kienstra commented:
Sounds good, Luke! Thanks for all of your work on this.

RobStino commented:
Works perfectly!

**Form**
![image](https://user-images.githubusercontent.com/1578148/67647332-a8d0df80-f97d-11e9-80ed-2ce6d24e973d.png)

**Preview**
![image](https://user-images.githubusercontent.com/1578148/67647342-b1c1b100-f97d-11e9-9b59-a02961b1fd26.png)

**Front end**
![image](https://user-images.githubusercontent.com/1578148/67647360-c69e4480-f97d-11e9-8160-998578fe1ba5.png)

kienstra commented:
Wow, great! Thanks a lot for testing this.
